### DISCUSSION TITLE
How can TileMaker be used for generating contour lines tiles?

### DISCUSSION THREAD

**Discussion by adibarda** [systemed/tilemaker#493] [Q&A] [answered]
Hello,
Does anyone knows what needs to be added to the Lua script in order to add data necessary for rendering contour lines?
Given that I have these MapBox layers for the contour lines:

    {
      "id": "contour-line",
      "type": "line",
      "metadata": {"IHM:overlay": true},
      "source": "Contour",
      "source-layer": "contour",
      "paint": {
        "line-color": "rgba(161, 127, 86, 1)",
        "line-width": ["match", ["get", "rank"], [1, 2], 1, 2]
      }
    },
    {
      "id": "contour-label",
      "type": "symbol",
      "metadata": {"IHM:overlay": true},
      "source": "Contour",
      "source-layer": "contour",
      "layout": {
        "text-field": "{height}",
        "symbol-placement": "line",
        "text-font": ["Open Sans Cond Bold"],
        "text-size": 12,
        "symbol-avoid-edges": true,
        "text-max-angle": 45,
        "text-letter-spacing": 0.15,
        "text-keep-upright": false
      },
      "paint": {
        "text-color": "rgba(161, 127, 86, 1)",
        "text-halo-color": "rgba(255, 255, 255,0.8)",
        "text-halo-width": 1.5,
        "text-halo-blur": 1.5
      }
    }
    
Thanks!
Adi

**Comment by systemed** [comment, marked as answer]
It's very doable (and I've done it) but it's a multistep process and you'll need some other tools.

First get your elevation data - usually SRTM tiles.

Use gdal_contour (from the GDAL suite) to make contour shapefiles out of that.

Then write a very brief Lua script - it doesn't need to do much, just set the `rank` attribute accordingly in `shape_function`. (I don't know what your Mapbox style expects here, sorry.)

Run tilemaker with this Lua script, a .json pointing to the shapefiles, and the bounding box you want.

One thing to watch - shapefiles have a maximum size of 2GB, so if the area you're generating is larger than that, you'll need to split it up into regions and then merge the results. You can use tilemaker's `--merge` flag to merge a bunch of mbtiles, or alternatively `tile-join` from Tippecanoe.

**Reply by adibarda** [reply]
Thank you for your answer. In the documentation it states that the Lua script support those functions:

  **node_function**(node), a function to process an OSM node and add it to layers
  **way_function**(way), a function to process an OSM way and add it to layers

so which one of those is to be used when you mention using **shape_function**? or is there another "shape_function" which is not mentions in the docs?
and as for the "rank" attribute - did you mean the "ele" (or "ELE") attribute which states the height of a given line?
If it's possible, I will appreciate a sample Lua script code. 
Thanks again.

**Comment by systemed** [comment]
Sorry, it's `attribute_function` rather than `shape_function`. See the "Shapefiles" section of CONFIGURATION.md.

Your style appears to read the `rank` attribute from the vector tiles (see line 8). So your `attribute_function` needs to set that attribute based on the raw `ele` value from your shapefiles.

**Reply by adibarda** [reply]
It works! I can see the contour lines in my map but I have a few questions:

1. Does the base PBF file mandatory? I mean if I just want to create MBTILES file of the contour lines is there a way not to provide the base PBF map and just use the SHP file? currently what I did the achieve this was to provide a configuration & Lua script with just the parts of the contour layer - no way_func / node_func etc. But still I provided the PBF map itself as an input to TileMaker

2.  The SHP file is very big (~1 Gig) for a relatively small area, yielding a relatively big ~190MB MBTILES file. I'm using this GDAL command:  gdal_contour -b 1 -a ELEV -i 10.0 -f "ESRI Shapefile" isr_geotiff.tif isr-contour.shp 
In order to make it smaller, should I try to optimize the gdal_contour command or manipulate TileMaker's attribute_function?

3.   I can see on the map a lot of unnecessary contour lines (small isolated polygons). Is there a way to reduce them using the attribute_function or should I look again for GDAL optimization?

Thanks a lot!

**Comment by systemed** [comment]
A .pbf isn't necessary but, if you omit one, you'll need to provide the bounding box on the command line.

To minimise the size, there's a few options you can look at (see CONFIGURATION.md):

- Use the layer definitions in your config.json to set simplification for lower zoom levels
- Use `_minzoom` in `attribute_function` to limit which contour lines are written at which zoom levels (for example, you could show them at 10m intervals at z14, 50m intervals at z13, 100m intervals at z12, etc.)
- Tweak your GDAL settings

> Is there a way to reduce them using the attribute_function or should I look again for GDAL optimization?

tilemaker doesn't offer a way to do this - it's something I'd like to do in the future but don't hold your breath. :)

**Comment by adibarda** [comment]
So, FinalyI have a beautiful contour lines in my Nav App. Thanks for your support!
If anyone sees this thread here are the configurations that worked for me. Hopefully it will help you too:

**Lua Script:**
========
`
-- Enter/exit Tilemaker
function init_function()
end
function exit_function()
end

-- Process node tags

node_keys = { "addr:housenumber","aerialway","aeroway","amenity","barrier","highway","historic","leisure","natural","office","place","railway","shop","sport","tourism","waterway" }
function node_function(node)
end

-- Process way tags

function way_function(way)

end

function attribute_function(attr,layer)

	local retval = {}
	
	if (attr["ELEV"]%50==0) then
		retval["rank"] = 3
		retval["height"] = attr["ELEV"]
	else 
		retval["rank"] = 1
	end
	
	return retval;

end`

**Config JSON:**
==========
`{
	"layers": {
		
		"contour": {"minzoom": 12, "maxzoom": 14, 
		            "source": "isr-contour.shp", 
					"simplify_below": 13, "simplify_level": 0.0003,
					"source_columns": ["ELEV"]}
 
	},
	"settings": {
		"minzoom": 12,
		"maxzoom": 14,
		"basezoom": 14,
		"bounding_box_xxxxxxx": [28.564,28.120,41.594,34.597],
		"include_ids": false,
		"combine_below": 14,
		"name": "Tilemaker to OpenMapTiles schema",
		"version": "3.0",
		"description": "Tile config based on OpenMapTiles schema",
		"compress": "none",
		"filemetadata": {
			"tilejson": "2.0.0", 
			"scheme": "xyz", 
			"type": "baselayer", 
			"format": "pbf", 
            "tiles": ["https://advrider.co.il/data/{z}/{x}/{y}.pbf"]
		}
	}
}`

**MapBox Style (only the contour lines part):**
============================
`{
      "id": "contour-line",
      "type": "line",
      "metadata": {"IHM:overlay": true},
      "source": "Contour",
      "source-layer": "contour",
      "paint": {
        "line-color": "rgba(161, 127, 86, 1)",
        "line-width": ["match", ["get", "rank"], [1, 2], 0.5, 2]
      }
    },

{
      "id": "contour-label",
      "type": "symbol",
      "metadata": {"IHM:overlay": true},
      "source": "Contour",
      "source-layer": "contour",
      "filter": ["!in", "rank", 1, 2],
      "layout": {
        "symbol-spacing": 500,
        "text-field": "{height}",
        "symbol-placement": "line",
        "text-font": ["Open Sans Cond Bold"],
        "text-size": 16,
        "symbol-avoid-edges": true,
        "text-max-angle": 45,
        "text-letter-spacing": 0.15,
        "text-keep-upright": false
      },

`

*Note - the SRTM data was downloaded & converted to shape file using QGIS
![image](https://github.com/systemed/tilemaker/assets/19632660/1c774ac0-aa20-451d-a45f-8f4097e1445b)

**Reply by systemed** [reply]
That looks awesome!

**Reply by WingmanImd** [reply]
Thanks @adibarda @systemed for the explanation and example code. One more thing that would be nice is to somehow get the lancover  attribute from the osm pbf file and set it as an attribute to the contour. That would allow for example to have different color contours and labels for forests, glacier and land. I've read in the docs about [spatial queries](https://github.com/systemed/tilemaker/blob/master/docs/CONFIGURATION.md#lua-spatial-queries) but that seems to be helpful for the reverse process, to add data to the tiles from, in this case 'contours'. Is there a way to add data to the contours from the osm pbf?

**Reply by systemed** [reply]
There isn't, I'm afraid - the shapefiles are read before the .osm.pbf is, so there's no way for the shapefiles to get information from the pbf.

**Reply by WingmanImd** [reply]
Thought that's the case. I'll find another way to "intersect" this data. Thank you!
